### DISCUSSION TITLE
FE 테스트 돌아보기 토론 | 2022.05.20

### DISCUSSION THREAD

**Discussion by woowapark** [woowacourse/retrospective#14] [General]
### 1 장바구니 스텝 2에서 단위 테스트를 어디에 적용했나요? 
- 그렇게 한 이유와 경험을 나눠주세요
- Redux와 별개로도 생각해보세요 🙂

### 2 내가 프로젝트를 진행하면서 테스트 전략을 세운다면, 어떤 범위의 테스트를 작성할 것 같은가요? 
- 단위 테스트, E2E 테스트, 시각적 테스트(ex. Storybook) 등 지금까지 경험해본 테스트 중 의미있게 느낀 경험이 있다면 공유해주세요 
- 내가 프로젝트를 진행할 때 어떤 테스트(들)을 선택할 것 같은가요?

### +) (있다면) 질문을 남겨주세요 
- 다른 크루들의 생각이 궁금한 질문이 있다면 남겨주세요
- 그룹에서 해결하지 못한 질문이 있다면 같이 남겨주세요

## 토론 페어
- 태태 & 소피아 + 안 & 코카콜라 
- 록바 & 샐리 + 나인 & 도리 
- 빅터 & 우연 + 콤피 & 유세지 
- 동키콩 & 꼬재 + 코이 & 온스타 
- 호프 & 마르코 + 앨버 & 자스민 
- 하리 & 우디 + 티거 & 밧드 
- 돔하디 & 후이 + 블링 & 준찌 
- 시지프 & 비녀 + 해리 & 무비 
- 병민 & 민초 + 결 & 위니 & 아놀드  

## 체크사항
✅ '어떤 생각과 고민의 과정을 거쳤는 지'가 잘 드러날 수 있게 이야기를 나누고 글로도 적어주세요. 

## 토론 나누기  
- 자기 그룹에서의 인사이트를 나눠주는 크루
- 다른 그룹의 질문에 기막히게 대답해주는 크루

**Comment by KangYunHo1221** [comment]
- 시지프
장바구니 자체를 테스트 할듯 하다. 단위테스트와 스토리북 테스트를 하면 좋을 것 같다.
스토리북을 하다보면 컴포넌트 기반으로 생각할 수 있게 됨.

- 비녀
단위테스트를 통해서 리팩토링 후 오류를 잡아냈던 기억이 있어서 하면 좋은 것 같다.
e2e하면 멋있어서 하고 싶다. 
백엔드 한테 보여주면 바로 인싸 가능할 것 같다.

- 무비
e2e를 하면서 손으로 테스트 해야됐던 기본사항들은 알아서 체크해줘서 좋았던 경험이 있다.
단위 테스트와 e2e는 진행하는 것이 좋지 않을까 싶다.

- 해리
테스트를 하면서 코드가 더 나아진다고 생각함. 단위테스트를 하면서 관심사 분리를 할 수 있게 된다.
하지만 단위테스트만 했을 때 실제 서비스에서 생길 수 있는 오류와는 갭이 있어서 e2e도 같이 해주는게 좋을 듯
테스트는 유지보수에 도움을 주는 것! 개발단계에서는 필수적이지 않을 것 같아서, 다 구현한 후 작성할 것 같다.


------------------------------
## 같이 논의하고 싶은 점

- storybook에서 어느정도의 coverage를 가지고 test 해야 할지?
- 어느 타이밍에 test를 작성해야 하는가(개발 완료 후? 개발중 CDD? 등등)
- 스토리북에서 상호작용 테스트를 안하면 무의미 한가?
- CDD를 위해서 스토리북을 쓰는게 맞나?

**Comment by usageness** [comment]
### 콤피

우선 순위 : E2E ≥ 스토리북 > 단위 테스트

- 프로젝트 초반 혹은 규모가 작은 프로젝트라면 E2E와 Storybook으로만 테스트를 할 것이다.
    - 테스트 코드를 작성하는 것도 결국 시간이 소모되는 것이기에 초반에는 단위 테스트의 중요성이 떨어진다 생각.
    - 초기 기능 구현 시 E2E를 작성함으로서 개발 시 편리함을 느꼈었고, E2E 테스트 코드 자체는 작성에 많은 시간이 소모되지 않아, 작성 비용에 대비해서 얻는 이점이 더 크다 느꼈음.
- 단위 테스트는 초기 프로젝트 상태에서는 크게 필요성을 못 느꼈다. 단위 테스트는 기능의 유지/보수 혹은 기능이 확장될 때 필요하다 생각.
- 단위 테스트에 대한 생각.
    - 단위 테스트는 개발 명세가 될 수 있기에 로직이 복잡하거나, 네이밍만으로 의도 전달이 잘 되지 않을 때 사용할 것 같음.
    - 결제 또는 쿠폰 등, 버그가 발생 되었을 때 타격이 큰 중요한 기능에서는 필수적으로 구현이 필요할 것 같음.

### 우연

- E2E 테스트와 storybook 테스트는 반드시 할 것이다.
    - storybook의 경우 작은 컴포넌트를 정확히 만드는 것이 의미있다.
    - 컴포넌트를 확인할 때 따로 작업할 필요없이 storybook으로 UI를 바로 확인할 수 있어서 편했다.
    - E2E 테스트는 실행하면 자동으로 원하는 동작과 결과를 테스트할 수 있어서 편리했다.
- 단위 테스트
    - 단위 테스트의 경우 프로젝트의 크기에 따라 필요성이 달라질 것 같다.
    - 큰 프로젝트일 경우 로직들이 복잡하게 얽힐 수 있고, 중요한 로직의 경우 단위 테스트가 필수적이다.
    - 작은 프로젝트의 경우 E2E, storybook으로 충분할 것 같다.
    - 로직의 flow를 이해하기 어려운 경우 단위 테스트로 코드의 flow를 확인할 수 있다.

### 유세지

**어떤 테스트를 적용할까?**

- 중요도 : E2E -> UI (Storybook) -> Unit
- 프론트엔드 특성상 UI 테스트 또한 의미가 크다.

**테스트를 왜 해야할까?**

- 테스트는 단순히 테스트 만을 위해 존재하지 않는다.
- 컴포넌트 단위 테스트하기 힘든 이유 : UI와 로직이 너무 결합되어있다.
    
    ⇒ 결합성이 강할수록 확장과 재사용이 어려워진다.
    
- 단위 테스트는 UI와 로직이 잘 분리됐는지 확인할 수 있는 수단이다.
    
    ⇒ 단위 테스트 작성이 힘들다면 UI와 로직이 밀접하게 결합됐다는 신호일 수 있다.
    
- 결국 단위 테스트는 좋은 코드를 만들기 위한 과정이다.

**함께 고민하고 싶은 주제**

철저한 테스트가 효율적인 시점은 프로젝트가 어느정도 개발된 이후인데, 빠르게 찍어내야하는 상황(시장의 반응을 파악하기 위해)에서도 테스트를 작성해야 할까요?

### 빅터

- 우선순위 : 스토리북 > E2E > 단위 테스트
- 스토리북
    - 작은 컴포넌트 단위로 상호작용하면서 빠르게 테스트 할 수 있는 장점이 있다.
    - UI 상 엣지 케이스를 테스트하기 용이하다.(ex. 버튼 내부 테스트를 길게 입력했을 때 어떻게 보이는 지에 관한 테스트)
- E2E
    - 플로우를 테스트하기 용이하다.
    - 사용자 플로우가 복잡하다면 반드시 필요하다고 생각한다.
- 단위 테스트
    - 로직을 테스트할 때 적합하다.
    - 중요한 로직(서비스의 핵심 로직 또는 결제 관련)에는 단위 테스트가 필요하다고 생각한다.
    - 단위 테스트 코드가 잘 작성돼 있으면 협업하는 팀원이 테스트 코드만 보고 로직을 파악하기 용이할 것이다. 따라서 복잡한 로직에 단위 테스트를 적용하는 것이 적합할 것 같다.
    - 프로젝트의 규모가 작다면 로직이 비교적 단순할 것이라 생각한다. 따라서 이런 경우 앞선 두 테스트에 비해 중요도가 떨어진다고 생각한다.

**Comment by soyi47** [comment]
## 태태 & 소피아 X 안 & 코카콜라

### 1. 장바구니 스텝 2에서 단위 테스트를 어디에 적용했나요?
- 다들 아직 테스트를 적용하지 못 했다 😥

### 2. 내가 프로젝트를 진행하면서 테스트 전략을 세운다면, 어떤 범위의 테스트를 작성할 것 같은가요?
- 단위 테스트, E2E 테스트, 시각적 테스트(ex. Storybook) 등 지금까지 경험해본 테스트 중 의미있게 느낀 경험
    - 작은 단위의 UI 컴포넌트, 어디든 갖다 쓸 수 있는 컴포넌트를 만들어서 스토리북 테스트 해보는 건 좋은 것 같다.

- 프로젝트를 진행할 때 어떤 테스트를 선택할 것인가?
   - 어플리케이션의 고유한 기능에 대해서는 기능 단위 테스트들을 작성한다.
   - API가 없는데 빠르게 보여줘야 한다면, 시각적 테스트를 작성한다.
    - 크리티컬한 상호작용에 대해서는 E2E 테스트를 작성한다. E2E는 요구사항이 계속 바뀌면 작성해도 유지하기가 힘들 것 같다. 

   - API 명세서만 잘 정해져 있으면, mocking data, msw를 사용해서, 시안에 맞게 잘 나오는지 테스트한다. 
   - 디자이너와 기획자와 다함께 협업할때, 만들어둔 스토리북을 주면 링크타고 들어가서 바로 확인하도록 하고 수정사항을 전달 받으면 편할 것 같다.

### 3. 추가 질문
   - react에서의 테스트 단위를 어떻게 잡으면 좋을까? 컴포넌트? 컴포넌트 내부의 로직?
   - 장바구니 미션에서 핵심적으로 테스트가 필요한 부분들은 어디일까?

**Comment by liswktjs** [comment]
## 샐리

- 테스트를 작성해보고 싶은 로직들
    - api 상품 id에 맞춰서 상품 정보를 가져오는지
    - 실패케이스 :  db에 없는 상품 아이디로 접근했을 때의 에러가 어떻게 처리 되는지
- 스토리북
    - 아토믹하게 렌더링이 어떻게 되는지 보여줄 때 편리했다
    - 협업할때  분리된 컴포넌트들을 한 번에 보여줄 때, 컴포넌트 분리 단위들을 시각적으로 보여줄 수 있어서 소통하기 편할 것 같다
    - 인터렉션하게 소통할 수 있는 스토리북을 구현한다면 더 의미가 있을 것 같다

## 도리

- E2E는 기능 구현을 확인할 때 사용하기 좋음. 사용자 플로우를 확인할 때 사용하면 된다. 나머지 간단한 로직은 단위 테스트를 사용할 생각이다. 시각적 테스트는 작은 컴포넌트를 확인할 때 사용하고, 이후에 E2E 이전에 다시 한 번 사용하면 좋을것 같다.

## 나인

- 정상 동작에 대한 테스트보다는 **에러 케이스에 대한 테스트**를 시도하려고 노력하자
    - 유효하지 않은 id에 접근했을때? 이런 느낌
- 스토리북 : atomic하게 짜면 좋은데 아직 실력이 의미있는 시각적 테스트까지는 안될 것 같다. (노력해보자)

- E2E : 익숙해서 가장 편한 테스트가 될 것 같음. 가장 만들기 쉽다고 생각
- Domain : 장바구니에서 도메인 로직이 많아서 가장 많이 시도해볼 것 같음.
- Storybook : 일단 멈춰!


## 록바

- 최대한 테스트 코드를 많이 작성하는 것이 좋은 것 같다.
- 테스트코드는 usecase에 대한 명세다.
- 주요 로직은 E2E테스트로,  나머지는 로직은 단위테스트 인터렉션이 있는 부분과 주요 공통컴포넌트는 스토리북으로 할 예정이다.

**Comment by dayelop** [comment]
## 🥕 하리 X 우디 X 밧드 X 티거

### ****장바구니 스텝 2에서 단위 테스트를 어디에 적용했나요?****

- 리덕스 리듀서 테스트
- api 요청에 대한 응답 테스트, 에러 응답 체크
- thunk에서 await가 되지 않음
- 렌더링 테스트를 하고 싶었지만 styled component에서 에러가 발생했고 해결하지 못 했음

### 내가 프로젝트를 진행하면서 테스트 전략을 세운다면, 어떤 범위의 테스트를 작성할 것 같은가요?

- 단위 테스트랑 스토리북 테스트는 꼭 하고 싶음
- UI는 핵심적인 부분에 대해서만 테스트
- TDD 어려움 ㅋㅋ (사실 안 어려움)
- 렌더링 테스트 왜 하는지 모르겠다
- 협업하는 개발자 입장에서 단위 테스트가 필요하다
- 스토리북 상호 작용 테스트 굿
- e2e 테스트는 개발 끝나고 테스트 용으로만 사용했다
- 단위 테스트 어디에서 했었는지 기억 안 나
- e2e 테스트가 수동 테스트 하는 시간을 줄여줘서 좋았음
- 스토리북으로 컴포넌트 하나 하나 볼 수 있어서 좋았다 CDD에 굿

### 질문

- 리액트 jest await이 안 먹어요 🥲
- 컴포넌트 렌더링 테스트에서 styled component 에러가 생겨요…

**Comment by kkojae91** [comment]
# 코이, 온스타 & 꼬재, 동키콩

## unit, e2e, storybook 중 의미 있는 경험 이야기


### 온스타

- 효용 있음 : FE관점 storybook 이 가장 효용 있었다.
	- 이유: 개발의 편의성 보장(컴포넌트 단위로 개발할 시) 내가 어떤 컴포넌트들을 만들었는지 확인할 수 있었다.

- 효용 없음 : 사이프레스
	- 이유: 작성하기 까다롭다...! 성공했을 때 상황 정의가 명확하지 않다..!
	- 생각해보니 효용이 있었던 것 같다...! 까다롭고 반복적인 상황을 테스트해야하는 경우!!!

- 프로젝트 진행시 테스트 어떤것 진행?
	- 스토리북 진행 


### 동키콩

테스트는 가치가 있지만, 기회비용에 있어서 분배를 잘해야할 것 같다.
- 테스트 해야하는 것과 하지 않아야 하는 것을 잘 구분하는게 중요한 것 같다.
- 구분의 기준은 명확하지않지만, 앞으로 구분에 대한 기준을 잘 선정하는게 중요한 것 같다.

프로젝트에서 어떤 테스트를 진행할 것 같은가?
- E2E는 하지 않을 것 같다..! (페어가 하자고 하면 의견 조율할 것 같다.)


### 코이

사이프레스는 온스타와 비슷한 생각을 가지고 있다. 뭔가 성공하지만 이렇게 테스트를 하는 방식이 맞나...?? 명확하지 않은데..?!

단위, 스토리북 -> 그래도 의미가 있다고 느낀다.
스토리북 : 컴포넌트를 작게 만들면서 바로 확인할 수 있어서 좋다.(import 불필요 하니까..?!)
- 하지만 그게 아니라면 굳이...?? 

단위테스트 : 도메인을 설계할 때 의미가 있는 것 같다.

프로젝트에서 어떤 테스트를 진행할 것 같은가?
- 단위 테스트만 진행


### 꼬재

동키콩이랑 생각이 비슷한 것 같다.
테스트는 효용이 있다고 생각하지만, 기회 비용이 있어 해야하는 테스트와 하지말아야할 테스트를 구분하는게 중요한 것 같다.

프론트는 UI가 자주 변경되니까 테스트를 진행하는 의미가 있는지는 공감하지 못했지만, 아직 경험하지 못한 부분이다 보니 아직은 테스트가 필요해 보인다.

프로젝트에서 어떤 테스트를 진행할 것 같은가?
- 시간이 있다면 세가지 모두 진행하고 싶다.


## 공통의견

### 스토리북
- 디자이너와 협업할 때 좋은 것 같다.

### E2E
- 까다롭고 반복되는 작업은 E2E 테스트로 작성해두고 확인하는게 좋은 것 같다.

### 단위
- ???

**Comment by airman5573** [comment]
## 1 장바구니 스텝 2에서 단위 테스트를 어디에 적용했나요?

### 민초(예정)
	- 장바구니에 추가 버튼을 눌렀을때 장바구니에 들어가는지
	- 수량을 올리고 내리고 차가 하거나 가각의 액션이 정확하게 동작 하는지

### 병민(예정)
	- 장바구니만
		- state 체크
		- dispatch
		- 다시 state체크
### 위니(예정)
	- 장바구니 추가 삭제
		- dispatch 날리면
		- 액션 딱 나타나고
		- 그 액션에 따라서 store가 바뀌는지
### A놀드
	- 장바구니
		- 체크 박스를 눌렀을때 내가 원하는 액션이 디스패치 된다
		- 스토어가 내가 생각했던 방식으로 바뀌는지
		- 스토어 상태에 따라서 내가 생각했던 뷰대로 바뀌는지
		- 상태에 따른 뷰가 제대로 반영이 되는지
### 결
	- 액션에 대한 테스트를 했습니다.

**Reply by rladpwl0512** [reply]
## 2 내가 프로젝트를 진행하면서 테스트 전략을 세운다면, 어떤 범위의 테스트를 작성할 것 같은가요?

공통적으로 E2E 테스트는 해야 할 의미를 잘 모르겠다는 의견이 있었습니다.
### 아놀드
storybook 세모(현재는 의미가 있는지 잘 모르겠다) / 단위 테스트 O(store, reducer) / E2E 테스트 X
### 결
storybook O (시각적 테스트) / 단위 테스트 O (중요한 로직)
### 민초
storybook O / 단위 테스트 O (리팩터링할 때 좋았다) / E2E 테스트 X
### 병민
storybook O (잘 쓸 것 같다.큰 단위 말고 작은 단위로, 공통으로 쓰이는 것들) / 단위 테스트 O / E2E 테스트 X
### 위니 
storybook O (bottom-up 단위 개발할 때 시각적 테스트가 좋았다) / 단위 테스트 O / E2E 테스트 X

**Comment by kwannee** [comment]
### 블링

1. 어디에 적용했나요?
    1. 리듀서와 액션에만 적용했다.
        1. 가장 확신이 없던 부분이 리덕스였기 때문에 로직이 불안했다. 또한 리덕스는 함수 단위이기 때문에 단위 테스트가 쉬웠다. 인풋과 아웃풋이 예측 가능했다.
    2. 왜 컴포넌트 테스트는 못했는가?
        1. 코드를 테스트하기 힘들게 짠 것 같다. 커스텀 훅을 사용했는데, 해당 훅을 포함한 모든 로직이 완벽하게 테스트하려면 불필요한 데이터까지 모두 mocking 해야 하는 어려움이 있었다. 

### 돔하디 & 후이 - 테스트 적용 못함

1. 테스트 전략을 세운다면, 어떤 범위의 테스트를 적용할 계획인가?
    1. 단위 테스트 → 로직이 복잡해지면 단위 테스트의 의미가 클 것 같은데, 현재는 리듀서의 리턴 값이 예상한 대로 나오는지 보는 정도다 보니 의미가 많이 줄어 들 것 같다.
    2. E2E → 그런 점에서 E2E의 효용은 좋은 것 같다. 도메인과 UI를 아우르게 테스트할 수 있었다. 환경 설정에 소모되는 시간도 적다는 이점이 있다.
    3. 스토리 북 → 개발 초기에 작은 단위의 컴포넌트를 확인하는데 좋다. 적용하지는 못했지만 로딩 상태, 에러 상태를 확인하는 데에 큰 도움이 됐을 것 같다. FE가 아닌 다른 분야의 팀원에게 현재 진행 상황을 실시간으로 보여줄 수 있는 장점도 있다.

**Comment by wonsss** [comment]
## 앨버
storybook을 활용해서 bottom-up 개발을 할 때 컴포넌트 별로 생각할 수 있어서 좋은 것 같다. 근데, 컴포넌트가 상위 컴포넌트의 스타일에 의존하게 되는 경우 decorator를 활용해서 테스트를 해주는데, 이는 추후 유지보수적으로 어려운 부분이 있는 것 같다

## 자스민

주문 목록을 테스트하였다. tdd를 하여서 테스트를 짜고 구현부분을 작성하였다.
프로젝트를 진행한다면?
스토리북 >> 단위테스트 > e2e
굳이 e2e정도는 스토리북의 상호작용 테스트로 커버할수 있을거 같다.
이중에 하나만 할수 있다면 스토리북 테스트를 할것 같다

## 호프

아직 적용은 못했지만, 생각해보자면 아래와 같다.
컴포넌트 테스트의 경우는 스토리북 / 상호작용 테스트만 해도 충분할 것 같다. (굳이 컴포넌트 렌더링 되는 것을 단위테스트로 작성하는 것이 유의미한지 모르겠다.)
그 외의 api 테스트 / 리듀서 테스트는 단위테스트를 작성 하여 비즈니스 로직이 정상동작하는지 확인 할 것 같다.

## 마르코

CDD를 통한 개발 단계에서 스토리북을 통해 시각 테스트하는 게 도움이 됐다. 테스트를 해야 한다면, 스토리북 및 단위테스틀 주로 사용할 것 같다. E2E는 스토리북 상호작용 테스트를 통해 어느 정도 커버할 수 있을 것 같다. 또한, 개발 단계에서 스토리북이 기획자나 디자이너 등과 협업에 도움이 될 것같다.
